File: gfx/shared/uniform.c
/*
 * Copyright (C) 2017 Manh Tran
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 */
#include <cherry/graphic/uniform.h>
#include <cherry/memory.h>
#include <cherry/array.h>
#include <cherry/bytes.h>

struct uniform_buffer *camera_uniform_buffer_alloc()
{
        struct uniform_buffer *p        = smalloc(sizeof(struct uniform_buffer));
        p->data                         = smalloc(sizeof(struct camera_uniform));
        p->types                        = smalloc(sizeof(union camera_uniform_type));
        p->update_indice                = array_alloc(sizeof(u8), UNORDERED);
        union camera_uniform_type *t    = p->types;
        t->project.mask         = -1;
        t->project.type         = UNIFORM_M4;
        t->project.offset       = offsetof(struct camera_uniform, project);
        t->project.frame        = 0;
        t->view.mask            = -1;
        t->view.type            = UNIFORM_M4;
        t->view.offset          = offsetof(struct camera_uniform, view);
        t->view.frame           = 0;
        t->position.mask        = -1;
        t->position.type        = UNIFORM_F3;
        t->position.offset      = offsetof(struct camera_uniform, position);
        t->position.frame       = 0;
        uniform_buffer_init(p, sizeof(struct camera_uniform));
        return p;
}

struct uniform_buffer *transform_uniform_buffer_alloc()
{
        struct uniform_buffer *p        = smalloc(sizeof(struct uniform_buffer));
        p->data                         = smalloc(sizeof(struct transform_uniform));
        p->types                        = smalloc(sizeof(union transform_uniform_type));
        p->update_indice                = array_alloc(sizeof(u8), UNORDERED);
        union transform_uniform_type *t = p->types;
        t->transform.mask               = -1;
        t->transform.type               = UNIFORM_M4;
        t->transform.offset             = offsetof(struct transform_uniform, transform);
        t->transform.frame              = 0;
        uniform_buffer_init(p, sizeof(struct transform_uniform));
        return p;
}

struct uniform_buffer *light_uniform_buffer_alloc()
{
        struct uniform_buffer *p        = smalloc(sizeof(struct uniform_buffer));
        // uniform_buffer_init(p, sizeof());
        return p;
}

void uniform_buffer_set(struct uniform_buffer *p, u8 index, void *data, size_t len)
{
        p->update                       = 1;
        struct uniform_type *types      = p->types;
        if(types[index].frame == 0) {
                array_push(p->update_indice, &index);
        }
        smemcpy(p->data + types[index].offset, data, len);
        types[index].frame              = BUFFERS;
        types[index].mask               = -1;
}

File: gfx/mtl/uniform.m
/*
 * Copyright (C) 2017 Manh Tran
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 */
#import <cherry/graphic/uniform.h>

#if GFX == MTL
#import <cherry/graphic/metal.h>
#import <cherry/memory.h>
#import <cherry/array.h>
#import <cherry/bytes.h>
#import <cherry/stdio.h>
#import <cherry/graphic/device_buffer.h>

void uniform_buffer_init(struct uniform_buffer *p, size_t size)
{
        int i;
        for_i(i, BUFFERS) {
                p->buffers[i] = device_buffer_alloc(BUFFER_VERTICE, size, BUFFER_PINNED);
                device_buffer_fill(p->buffers[i], p->data, size);
                p->buffers[i]->ref++;
        }
}

void uniform_buffer_update(struct uniform_buffer *p, u8 frame)
{
        struct uniform_type *types      = p->types;
        u8 *it;
        i16 it_index;
        array_for_each_index(it, it_index, p->update_indice) {
                u8 index                = *it;
                if(types[index].mask == frame) continue;

                types[index].frame--;
                types[index].mask       = frame;

                if(types[index].frame == 0) {
                        array_remove(p->update_indice, it_index);
                        it_index--;
                        it--;
                }

                void *data              = p->data + types[index].offset;
                switch (types[index].type) {
                        case UNIFORM_F1:
                                device_buffer_sub(p->buffers[frame], types[index].offset, data, sizeof(float));
                                break;
                        case UNIFORM_F2:
                                device_buffer_sub(p->buffers[frame], types[index].offset, data, sizeof(float) * 2);
                                break;
                        case UNIFORM_F3:
                                device_buffer_sub(p->buffers[frame], types[index].offset, data, sizeof(float) * 3);
                                break;
                        case UNIFORM_F4:
                                device_buffer_sub(p->buffers[frame], types[index].offset, data, sizeof(float) * 4);
                                break;
                        case UNIFORM_I1:
                                device_buffer_sub(p->buffers[frame], types[index].offset, data, sizeof(int));
                                break;
                        case UNIFORM_I2:
                                device_buffer_sub(p->buffers[frame], types[index].offset, data, sizeof(int) * 2);
                                break;
                        case UNIFORM_I3:
                                device_buffer_sub(p->buffers[frame], types[index].offset, data, sizeof(int) * 3);
                                break;
                        case UNIFORM_I4:
                                device_buffer_sub(p->buffers[frame], types[index].offset, data, sizeof(int) * 4);
                                break;
                        case UNIFORM_M2:
                                device_buffer_sub(p->buffers[frame], types[index].offset, data, sizeof(union mat2));
                                break;
                        case UNIFORM_M3:
                                device_buffer_sub(p->buffers[frame], types[index].offset, data, sizeof(union mat3));
                                break;
                        case UNIFORM_M4:
                                device_buffer_sub(p->buffers[frame], types[index].offset, data, sizeof(union mat4));
                                break;
                        default:
                                break;
                }
        }
}

void uniform_buffer_free(struct uniform_buffer *p)
{
        sfree(p->types);
        sfree(p->data);
        array_free(p->update_indice);
        int i;
        for_i(i, BUFFERS) {
                device_buffer_free(p->buffers[i]);
        }
        sfree(p);
}

#endif
